Fire diagnostic source events from IHostBuilder.Build

davidfowl opened the pull request:
- We want to enable getting access to the `IHostBuilder` and `IHost` during the call to `IHostBuilder.Build` so that we can access the application's `IServiceProvider` from various tools (like the EntityFramework migrations tool or scaffolding). Usually, this is enabled by exposing a different entry point from the application (`CreateHostBuilder`) but this technique allows us to let the application execute normally and hook important events via a diagnostic source.
- Updated HostFactoryResolver and added new APIs for resolving the IHostBuilder from an assembly with an entry point. 
- Added tests

~TODO: I need to update the `HostFactoryResolver` to support this new pattern.~

~TODO: Add tests for various scenarios:~
- ~Exception thrown before IHostBuilder.Build is called where the exception propagates~
- ~Exception thrown before IHostBuilder.Build is called where the exception does not propagate~
- Crazy scenarios where multiple host builders are fired up at during the call to main (not sure this is important)

cc @halter73

davidfowl committed:
Fire diagnostic source events from IHostBuilder.Build
- We want to enable getting access to the IHostBuilder and IHost during the call to IHostBuilder.Build so that we can access the application's IServiceProvider from various tools. Usually, this is enabled by exposing a different entry point from the application but this technique allows us to let the application execute normally and hook important events via a diagnostic source.

davidfowl committed:
Handle linker warnings

davidfowl committed:
Remove DAM annotations

davidfowl committed:
Add support for the new pattern in HostFactoryResolver
- Added support for resolving an IServiceProvider using the new diagnostics source pattern (which doesn't require a different entrypoint)

davidfowl commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +139,158 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+
+            public void OnNext(DiagnosticListener value)
+            {
+                if (value.Name == "Microsoft.Extensions.Hosting")
```
This is where we wire up to the new event to intercept calls to build.

davidfowl commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +139,158 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+
+                // Wait before throwing an exception
+                using var cts = new CancellationTokenSource(_waitTimeout);
+                _hostTcs.Task.Wait(cts.Token);
```
I'm not sure if we should propagate the OperationCanceledException here or return null.

davidfowl commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +139,158 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+            private Action<IHostBuilder> _configure = b => { };
+
+            // The amount of time we wait for the diagnostic source events to fire
+            private static readonly TimeSpan _waitTimeout = TimeSpan.FromSeconds(5);
```
I'm not sure what a decent default should be. I had 30 seconds but that would make testing invalid tests kinda awful.

davidfowl committed:
Made fixes and added more tests

davidfowl committed:
Handle main with no args

davidfowl committed:
Make sure we copy properties over

davidfowl committed:
Store the entrypoint directly

davidfowl committed:
Added tests for top level statements

davidfowl committed:
Mark tests that require threading

davidfowl committed:
Run tests with RemoteExecutor since they rely on a global event
- DiagnosticSources will publish events globally and we need to make sure there's no cross talk between tests

davidfowl committed:
Increase the timeout

martincostello commented:
Would this light-up the ability to use `WebApplicationFactory<T>` (and maybe top-level programs) with the new minimal hosting APIs?

davidfowl commented:
Yes that's the intent:
- EF migrations
- WebApplicationFactory
- NSwag
- Scaffolding 

Etc

martincostello commented:
Nice 👍 - I've been holding off in using it in a bunch of stuff as so much of our tests use WebApplicationFactory to do integration-level tests.

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:
```diff
@@ -127,13 +129,40 @@ public IHost Build()
 
+            // REVIEW: If we want to raise more events outside of these calls then we will need to
+            // stash this in a field.
+            using var diagnosticListener = new DiagnosticListener("Microsoft.Extensions.Hosting");
```
I know we have the tests in the HostFactoryResolver, but it might be a good idea to add some tests directly to Extensions.Hosting that ensure this fires correctly. To prevent breaking it in the future.

davidfowl replied:
Done

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:
```diff
@@ -127,13 +129,40 @@ public IHost Build()
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:UnrecognizedReflectionPattern",
+            Justification = "The values being passed into Write are being consumed by the application already.")]
```
It isn't the "The values being passed into Write" that are the issue, but instead the recursive properties of the values.

DiagnosticSource has an EventSource bridge that allows end-users to set up "filter specs" that specify which properties of the objects passed here should be written to the EventSource. If those properties aren't used in the app, they could be trimmed.

davidfowl replied:
I think that's true in general but in the use case for these events, I'm not worried:
- There won't be any trimming happening in development. You don't publish an application trimmed and then run tools in it to grab the IHost and IHostBuilder.
- In the 90% case, the relevant properties/methods are going to be used as part of application bootstrapping.

eerhardt replied:
Of course, if we don't expect people to be using the EventSource bridge, that could be a "good enough" reason to suppress.

eerhardt replied:
> There won't be any trimming happening in development. You don't publish an application trimmed and then run tools in it to grab the IHost and IHostBuilder.

What about someone else who takes a dependency on this new DiagnosticSource?

davidfowl replied:
> Of course, if we don't expect people to be using the EventSource bridge, that could be a "good enough" reason to suppress.

Yes this isn't a use case. There's nothing serializable on there that can be sent out of process.

> What about someone else who takes a dependency on this new DiagnosticSource?

Then we'll learn about the scenario then because it won't work for that scenario we don't understand.

noahfalk replied:
I think anyone else who used the DiagnosticSource would either:
a) cast the received object to its correct type which would make the linker aware of the usage - everything works nicely here
b) private reflection - I trust this puts the onus on the new DiagnosticSource consumer to ensure the fields aren't removed

davidfowl committed:
Added test for events

eerhardt commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -31,6 +37,16 @@ internal sealed class HostFactoryResolver
             return ResolveFactory<THostBuilder>(assembly, CreateHostBuilder);
         }
 
+        public static Func<string[], IHostBuilder>? ResolveHostBuilderFactory(Assembly assembly)
```
1. Does this need to be public?
2. If it does, it is a bit concerning that it has the same name as the method above, takes the same parameters, and returns the same type, but doesn't really do the same things. Can we rename it somehow to distinguish it from the existing method above?

davidfowl replied:
I'll rename it.

eerhardt commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +139,184 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+            private Action<IHostBuilder> _configure;
+
+            // The amount of time we wait for the diagnostic source events to fire
+            private static readonly TimeSpan _waitTimeout = TimeSpan.FromSeconds(20);
```
```suggestion
            private static readonly TimeSpan s_waitTimeout = TimeSpan.FromSeconds(20);
```

eerhardt commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +139,184 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+            {
+                _args = args;
+                _entryPoint = entryPoint;
+                _configure = b =>
```
```suggestion
                _configure = innerBuilder =>
```

or something similar. To make it easier to understand what this variable is.

eerhardt commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +139,184 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+                }
+                catch (AggregateException) when (_hostTcs.Task.IsCompleted)
+                {
+                    // Lets this propogate out of the call to GetAwaiter().GetResult()
```
```suggestion
                    // Lets this propagate out of the call to GetAwaiter().GetResult()
```

eerhardt commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +139,184 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+                    // Wait before throwing an exception
+                    if (!_hostTcs.Task.Wait(_waitTimeout))
+                    {
+                        throw new InvalidOperationException("Unable to build IHost");
```
Maybe use a different message here - specifically call out that it timed out. That way we can tell the difference between the entrypoint returning gracefully vs. timing out.

davidfowl replied:
I want the timeout to be an implementation detail. I'm also thinking this could also return null.

noahfalk replied:
Who do you expect to see the message? If the message is primarily shown to people who are trying to resolve the issue (or users will relay it to someone else who resolves the issue) then it might help to make it more descriptive.

davidfowl replied:
I need to check the callers but its possible null is a better return than throwing.

eerhardt commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/CreateHostBuilderInvalidSignature/Program.cs:
```diff
@@ -2,14 +2,15 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using MockHostTypes;
+using Microsoft.Extensions.Hosting;
```
Are the changes to this file and `tests/CreateHostBuilderPatternTestSite/Program.cs` necessary?

davidfowl replied:
Because I added Hosting as a dependency yes.

eerhardt reviewed (approved):
Looks fine to me. Just some random nits.

bricelam reviewed (approved):
LGTM. I really like this approach--it's the magic I've been looking for since the beginning of ASP.NET Core.

As discussed offline, we should try hard to avoid the compile-time dependency on `Microsoft.Extensions.Hosting`. We had a lot of pushback against this in the 1.1 days since it adds it to all app models (see https://github.com/dotnet/efcore/issues/7835 from a WPF user for example)

/cc @ajcvickers

halter73 commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +139,184 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+                    // Lets this propogate out of the call to GetAwaiter().GetResult()
+                }
+
+                Debug.Assert(_hostTcs.Task.IsCompleted);
```
Nit: I think the when makes this clear enough

halter73 commented on src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:
```diff
@@ -127,13 +129,40 @@ public IHost Build()
+            var host = _appServices.GetRequiredService<IHost>();
+            if (diagnosticListener.IsEnabled() && diagnosticListener.IsEnabled(hostBuiltEventName))
+            {
+                Write(diagnosticListener, hostBuiltEventName, host);
```
Any particular reason for resolving the IHost before writing this event? I get that it resolves a bunch of services. Is that important for any particular reason? Why not leave that to the listener?

halter73 replied:
I guess the `_hostBuilt` flag being set would make it tricky to build in the subscriber once you get this far.

davidfowl committed:
Remove dependency on hosting in shared file

davidfowl committed:
Actually remove hosting dependency

davidfowl committed:
Detect the version of Microsoft.Extensions.Hosting before waiting for events to fire
- We want to fail fast if we know this version of hosting will never fire events of if the hosting assembly fails to load. Added a version check.
- Allow the caller to specify a wait timeout.

davidfowl committed:
Added a flag to allow running the application to completion

noahfalk commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -31,6 +40,35 @@ internal sealed class HostFactoryResolver
+            try
+            {
+                // Attempt to load hosting and check the version to make sure the events
+                // even have a change of firing (they were adding in .NET >= 6)
```
Nit:
```suggestion
                // even have a chance of firing (they were added in .NET >= 6)
```

noahfalk commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -112,5 +160,133 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
+                    if (_stopApplication)
+                    {
+                        // Stop the host from running further
+                        throw new StopTheHostException();
```
The prize for dirty hackery this week : D

Is it important to guarantee that the thread really does stop? It would be easy enough for an app to throw a try/catch around the part of their code where this gets raised so that it never gets back to the entrypoint. Your tool code will be running in parallel with the user's unknown error handling app code. It doesn't seem obviously harmful to me but figured I mention it.

davidfowl replied:
Disposal is one of the things I'm worried about and why I think throwing to stop execution makes the most sense here.

davidfowl replied:
> It would be easy enough for an app to throw a try/catch around the part of their code where this gets raised so that it never gets back to the entrypoint.

The thing that stops this from happening in the throw case is the fact that the application never gets access to the IHost instance. They can't call build again on it because double building throws. Even if they catch the exception and do something else, the IHost isntance that came out of Build is never observed by the application.

noahfalk replied:
Yeah, I wasn't imagining they continue normally, more like they have some complex error handling code that will be running in parallel. 

> the application never gets access to the IHost instance

[Joking] What do you mean? This PR just added the official mechanism that lets everyone access IHost without the pesky inconvenience of needing Build() to return it to you ;p

eerhardt commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -31,6 +40,35 @@ internal sealed class HostFactoryResolver
             return ResolveFactory<THostBuilder>(assembly, CreateHostBuilder);
         }
 
+        public static Func<string[], object>? ResolveHostFactory(Assembly assembly, TimeSpan? waitTimeout = null, bool stopApplication = true, Action<object>? configureHostBuilder = null)
```
I don't see a test that uses `configureHostBuilder`. Do we need this parameter?

davidfowl replied:
Yes, we will need it for https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs

eerhardt replied:
Ok, then it would be great if we had a test ensuring it doesn't break.

davidfowl replied:
2 things:
1. As of right now, ASP.NET Core will need to use it so it kinda *has* to work or tests will fail regardless 😄.
2. I agree with you and will add a test.

davidfowl commented:
@eerhardt and @noahfalk I need to get this merged so I can react in several places down stream. I will react to your feedback in another PR so keep commenting.
